pageserver/src/tenant.rs

```diff
@@ -1437,28 +1441,56 @@ impl Tenant {
         Ok(())
     }
 
-    /// Flush all in-memory data to disk.
+    /// Flush all in-memory data to disk and remote storage, if any.
     ///
     /// Used at graceful shutdown.
-    ///
-    pub async fn freeze_and_flush(&self) -> anyhow::Result<()> {
-        // Scan through the hashmap and collect a list of all the timelines,
-        // while holding the lock. Then drop the lock and actually perform the
-        // flushing. We don't want to block everything else while the
-        // flushing is performed.
-        let timelines_to_flush = {
+    async fn freeze_and_flush_on_shutdown(&self) {
+        let mut js = tokio::task::JoinSet::new();
+
+        // execute on each timeline on the JoinSet, join after.
+        let per_timeline = |timeline_id: TimelineId, timeline: Arc<Timeline>| {
+            async move {
+                debug_assert_current_span_has_tenant_and_timeline_id();
+
+                match timeline.freeze_and_flush().await {
+                    Ok(()) => {}
+                    Err(e) => {
+                        warn!("failed to freeze and flush: {e:#}");
+                        return;
+                    }
+                }
+
+                let res = if let Some(client) = timeline.remote_client.as_ref() {
+                    client.wait_completion().await
```

Comment:
It wasnt there before right? I'm concerned about potential increase in time it takes to shutdown (which creates same kind of issues as long startup)

Reply:
It was implicitly here outside of shutting down all tenants:

https://github.com/neondatabase/neon/blob/e0bd81ce1f9204a50a74c4f71baacee8220be5a6/pageserver/src/lib.rs#L61-L65

But I see your point ... In tests it was racy to just have `shutdown_tasks(None, Some(tenant_id), None);` because it the uploads may not have been started yet. So the upload tasks could be found to be shutdown in `shutdown_tasks(None, None, None);`

However as far as real effects towards shutdown time go, now everything is shutdown in parallel.

Unsure how to change this.

Reply:
Should add a comment for this.

Reply:
I think comment is enough. I'd say we should fix it when we're out of task_mgr restrictions

Reply:
Ok. Fixing this however is not trivial ... Yesterday I was thinking about does the queue thing even make any sense for RemoteTimelineClient, what would a race-less design look here.

Reply:
Thanks! Added this one https://github.com/neondatabase/neon/pull/4407#discussion_r1219544137